### ISSUE TITLE
A few questions on details

### ISSUE THREAD

**Issue by Divadi** [Nightmare-n/UniPAD#2]
Thank you for releasing this amazing work! I just had a couple questions on some of the camera-only outdoor details @Nightmare-n 

1. Are the same 6 images used for generating the 3D voxel grid and rendering (as is mentioned to be done for ScanNet in PonderV2)?
2. Were the used ConvNeXt(V1?) backbones trained from scratch or with IN1k?
3. Was any data augmentation used for the 2D training stage, besides regular MAE masking?
4. When using the proposed depth-aware sampling, are the same 512 rays, sampled from the pixels with available LiDAR points, used for both color & depth rendering?
5. In Table 8g, there appear to be trainable weights associated with the view transformation stage. Does the view transformation generally follow UVTR(?) with multi-scale sampling and depth weighting? Or perhaps single-scale?
6. In PonderV2, supplementary is mentioned. Would it be possible for this to be made public?
7. What (total) batch size was used?

Apologies for the list of questions, but I'm really interested in the work. Again, thank you so much in advance!

**Comment by Nightmare-n** [comment]
Hi, thanks for your interest ans question!
1. Yes, we use six multi-view images in nuScenes to generate 3D voxel grids.
2. We use ConvNeXt-V1 pre-trained on IN1k.
3. We only use MAE masking, as we do not observe obvious improvements when using other data augmentation like flip and resize.
4. Yes, we sample 512 rays for both color and depth rendering.
5. The view transformation contains multi-scale sampling and depth weighting, consistent with UVTR.
6. This seems to be a typo, which is copied directly from PonderV1.
7. We use 1 batch size per GPU and  4 GPUs for the camera-based model, which is the same as UVTR.

**Comment by Doctor-James** [comment]
Thank you for your work. I would like to ask how sampling 512 rays per image view achieves such a high rendering resolution in the final render image(figure3 in your paper). For a high-definition image of 1600 by 900, 512 seems like a very small number.

**Comment by Divadi** [comment]
@Nightmare-n Thank you so much for your earlier response. I think this is a strong work, so I've been trying to reproduce it! Unfortunately my results are not as good, and I wanted to get your guidance on some implementation details.

Specifically, I am trying to create this result:
![image](https://github.com/Nightmare-n/UniPAD/assets/17109672/af88c9c7-8978-4521-b745-189c430eabb8)

Here is what I've tried.

### Reproducing UVTR-C (Baseline)
First, I tried loading in ConvNeXt-S weights as in [mmdetection](https://github.com/open-mmlab/mmdetection/blob/main/configs/convnext/cascade-mask-rcnn_convnext-s-p4-w7_fpn_4conv1fc-giou_amp-ms-crop-3x_coco.py), which are from the official repo. Training for 12 epochs on load_interval=2, it achieves 20.9/22.4 for NDS/mAP. I tried removing augmentations and got 22.7/25.1, which is still lower NDS than shown. I based my code on [UVTR camera base](https://github.com/dvlab-research/UVTR/blob/main/projects/configs/uvtr/camera_based/uvtr_camera_base.py), adding layer-wise LR decay.

Next, I saw you reference [SparK](https://github.com/keyu-tian/SparK) pre-training, so I tried their pre-trained ConvNeXt-S weights. I also increased unified volume size to 180x180x5 (it's not mentioned what volume size you use for UVTR). This now gets 24.8/28.3 - NDS is still worse, but mAP is now too good...

Could I get guidance on the exact settings you used for this?

### Reproducing UniPAD
- I load in the SparK ConvNeXt-S checkpoint, then mask 0.3 and do 2D masked backbone as in SparK. 
- Then, aggregate the 6 image views into the 180x180x5 volume as in Camera-UVTR (no lidar info). 3 conv-bn-relu 3D CNNs, then CNN to 32 channels.
- For each image view, I sample 512 pixels with valid depth (from 10 past LiDAR sweeps).
- Ray sample 96 points uniformly from 1m to 64m. (No importance sampling)
- Grid sample features with align_corners, input into NeuS SDF with 10-freq XYZ (-1 ~ 1 norm) posemb, 6 layers, 32 hidden, get SDF + 32 feats.
- Put XYZ, volume feats, 32 SDF feats, 4-freq view, SDF gradient into 4-layer color network to get RGB.
- Render RGB & Depth, supervise w/ RGB weight 10, Depth weight 10. Notably, I just directly do:
![image](https://github.com/Nightmare-n/UniPAD/assets/17109672/14cf0ca0-adf7-4752-9ad8-699e3a16f8da)
between consecutive points and don't do the special NeuS gradient * viewdir [thing](https://github.com/Totoro97/NeuS/blob/629b9cf0d9e2451d10c498726b51a7940069ed13/models/renderer.py#L230-L235), because I felt that without eikonal loss, there are less guarantees on gradient magnitude.
- For points outside of volume, their density & RGB are just set to 0.

With this, training 6 epochs (I know this shouuld be 12, but for faster experiments) with load_interval=1, no augmentations, then fine-tuning UVTR (removing the 3 conv-bn-relu 3D CNNs), I get 26.9/31.7. Compared to 24.8/28.3 baseline, it's a +2.1/+3.4 improvement, but it's really far from the +7.7/+9.6 in paper. Notably, NDS doesn't improve too much...

EDIT: Trained for 12 epochs, achieves just 27.2/32.1

Do you have any ideas on what I can fix or what I may be missing?

Thank you so much in advance! If you are okay with having a more in-depth discussion, please feel free to email me directly as well.

**Comment by Nightmare-n** [comment]
Hi, the code is released!

**Comment by Sen-Tr** [comment]
Hi @Nightmare-n , thank you for wonderful work. I have tried to reproduce the results of UVTR-C + UniPAD (ablation study setup) given your released code but I couldn't get the final NDS as good as in your paper. 

[1]. I trained the model **abl_uvtr_cam_vs0.1_finetune.py** given the pretrained weights you released for UniPAD_C (voxel_size=0.1). After 12 epochs of finetuning (load_interval=2), I got:
_12 epochs:
mAP: 0.3124
mATE: 0.8361
mASE: 0.7168
mAOE: 1.5297
mAVE: 1.2606
mAAE: 0.3971
NDS: 0.2612 -> The NDS is pretty low_

[2]. I pretrained from scratch (12 epochs) and finetuned the model in the same manner using exactly the same code you released. I still got low NDS.
_12 epochs: loss=7.4540
mAP: 0.3204
mATE: 0.8247
mASE: 0.7141
mAOE: 1.5130
mAVE: 1.1473
mAAE: 0.3730
NDS: 0.2690 -> The NDS is pretty low_

When I evaluated your ckpt: **abl_uvtr_cam_vs0.1_finetune.pth**, I found that the mASE and mAOE of above reproduced results are noticeably higher than below:
_Evaluation of your provided ckpt:
mAP: 0.3285
mATE: 0.8140
mASE: 0.2912
mAOE: 0.8517
mAVE: 1.2607
mAAE: 0.3872
NDS: 0.3298_

Do you have any ideas about that? Could you please provide the log file recording your training process?
Thank your very much.
